src/rand.c

```diff
@@ -0,0 +1,231 @@
+/*  Written in 2018 by David Blackman and Sebastiano Vigna (vigna@acm.org)
+
+To the extent possible under law, the author has dedicated all copyright
+and related and neighboring rights to this software to the public domain
+worldwide. This software is distributed without any warranty.
+
+See <http://creativecommons.org/publicdomain/zero/1.0/>. */
+
+#include "common.h"
+#include "rand.h"
+#include "runtime.h"
+
+#if defined(GIT_RAND_GETENTROPY)
+# include <sys/random.h>
+#endif
+
+static uint64_t state[4];
+static git_mutex state_lock;
+
+/*
+ * On 32-bit systems, we want to put inputs to the seed in the high
+ * and low parts of the 64-bit seed, so we shift some into place.
+ */
+#ifdef GIT_ARCH_32
+# define SEED_SHIFT 32
+#else
+# define SEED_SHIFT 0
+#endif
+
+#if defined(GIT_WIN32)
+GIT_INLINE(int) getseed(uint64_t *seed)
+{
+	HCRYPTPROV provider;
+	SYSTEMTIME systemtime;
+	FILETIME filetime, idletime, kerneltime, usertime;
+
+	if (CryptAcquireContext(&provider, 0, 0, PROV_RSA_FULL,
+	                        CRYPT_VERIFYCONTEXT|CRYPT_SILENT)) {
+		BOOL success = CryptGenRandom(provider, sizeof(uint64_t), (void *)seed);
+		CryptReleaseContext(provider, 0);
+
+		if (success)
+			return 0;
+	}
+
+	GetSystemTime(&systemtime);
+	if (!SystemTimeToFileTime(&systemtime, &filetime)) {
+		git_error_set(GIT_ERROR_OS, "could not get time for random seed");
+		return -1;
+	}
+
+	/* Fall-through: generate a seed from the time and system state */
+	*seed = 0;
+	*seed |= ((uint64_t)filetime.dwLowDateTime << 32);
+	*seed |= ((uint64_t)filetime.dwHighDateTime);
+
+	*seed ^= ((uint64_t)GetCurrentProcessId() << 32);
+	*seed ^= ((uint64_t)GetCurrentThreadId());
+
+	GetSystemTimes(&idletime, &kerneltime, &usertime);
+	*seed |= ((uint64_t)idletime.dwLowDateTime);
+	*seed |= ((uint64_t)idletime.dwHighDateTime << 32);
+	*seed |= ((uint64_t)kerneltime.dwLowDateTime << 32);
+	*seed |= ((uint64_t)kerneltime.dwHighDateTime);
+	*seed |= ((uint64_t)usertime.dwLowDateTime);
+	*seed |= ((uint64_t)usertime.dwHighDateTime << 32);
```

Comment:
Is this supposed to be `|=` for these 6?
It "feels" like they were supposed to be `^=`

Reply:
Meh, I actually meant to compose the high and low bits into a single 64 bit number and then xor _that_.